src/core/experimental/buffer-map.js

```diff
@@ -0,0 +1,165 @@
+import GL from '../../webgl/gl-constants';
+import Buffer from '../../webgl/buffer';
+import TransformFeedback from '../../webgl/transform-feedback';
+import Model from '../model';
+import {isWebGL2, assertWebGL2Context} from '../../webgl/context';
+import assert from 'assert';
+
+const PASS_THROUGH_FS = `\
+void main()
+{
+}`;
+
+export default class BufferMap {
+
+  static isSupported(gl) {
+    // For now WebGL2 only
+    return isWebGL2(gl);
+  }
+
+  constructor(gl, opts = {}) {
+    assertWebGL2Context(gl);
+
+    this.gl = gl;
+    this.model = null;
+    this._buffersSwapable = false;
+    this.currentIndex = 0;
+    this.sourceBuffers = new Array(2);
+    this.destinationBuffers = new Array(2);
+    this.transformFeedbacks = new Array(2);
+    this._buffersToDelete = [];
+    this.elementCount = 0;
+
+    this.initialize(opts);
+    Object.seal(this);
+  }
+
+  initialize({
+    sourceBuffers = null,
+    destinationBuffers = null,
+    vs = null,
+    sourceDestinationMap = null,
+    varyings = null,
+    drawMode = GL.POINTS,
+    elementCount = null
+  }) {
+    assert(sourceBuffers && vs && varyings && elementCount);
+
+    // If destinationBuffers are not provided, sourceDestinationMap must be provided
+    // to create destinaitonBuffers with layout of corresponding source buffer.
+    assert(destinationBuffers || sourceDestinationMap);
+
+    this.bindBuffers({sourceBuffers, destinationBuffers, sourceDestinationMap});
+
+    let index = 0;
+    this.varyings = [];
+    this.varyingMap = {};
+    for (const varying of varyings) {
+      this.varyings[index] = varying;
+      this.varyingMap[varying] = index;
+      index++;
+    }
+
+    for (let i = 0; i < 2; i++) {
+      this.transformFeedbacks[i] = new TransformFeedback(this.gl, {
+        buffers: this.destinationBuffers[i],
+        varyingMap: this.varyingMap
+      });
+    }
+
+    // Append matching version string to FS.
+    let fs = PASS_THROUGH_FS;
```

Comment:
`Model` object takes shader Strings as input, at this point we don't have access to `Shader` unless we create them and pass it to Model instead of Strings. But I like the idea of managing `version` string inside Shader, lets explore this idea in a separate PR.